### PR TITLE
[Orders with Coupons M7]  Add informational tooltips

### LINKED ISSUE #10839: [Orders with Coupons M7] Add informational tooltips

### PR DESCRIPTION
**Pull request by iamgabrielma**
Closes: #10839. 
#10837 should be merged first

## Description
This PR adds informational tooltips on tapping the `(?)` symbol when adding coupons to an order, or when adding discounts to a product. This PR deals with the UI only, in the following we'll deal with the logic that makes coupons and discounts mutually exclusive.

| Orders | Orders tooltip |
|--------|--------|
| _coupons_tooltip | Simulator Screenshot - iPhone 15 - 2023-10-10 at 20 19 42 | 

| Product | Product tooltip |
|--------|--------|
| _discounts_tooltip  | Simulator Screenshot - iPhone 15 - 2023-10-10 at 20 20 02 | 

## Changes:
* Extracts the product discount row into a `@ViewBuilder` variable for cleaning up the main view. We also move all other `@ViewBuilder` vars to a private extension. 
* Adds `shouldShowCouponsInfoTooltip` state which will change when the `(?)` icon is tapped. This will show/hide an overlay with the informational tooltips.

## Testing instructions
* Switch `ordersWithCouponsM6` feature flag to true
* Go to `Orders` > `+` > Observe that:
  * `+ Add Coupon` is disabled. `(?)` is tappable. Tapping on it toggles the informational tooltip. Tapping again hides it.
  * Tap on `+ Add Product` and add a product to the order. Now `+ Add Coupon` is enabled, and the `(?)` icon is hidden.
  * On the product you've added to the order, or a new product added to the order, tap the chevron to expand the collapsable product row.
  * Observe that the `(?)` in `+ Add discount` is tappable, tapping it shows an informational tooltip and tapping again hides it.
  * Adding a discount to the product will hide both the `+ Add discount` and the informational tooltip icon.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/OrderPaymentSection.swift [hunk 1]
```diff
@@ -82,35 +88,36 @@ struct OrderPaymentSection: View {
             }
 
             addCouponRow
+                .zIndex(1)
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/OrderPaymentSection.swift:98]: This is an interesting one: My understanding is that you'd need a ZStack in order to apply `.zIndex`, but that [doesn't seem to be the case](https://developer.apple.com/documentation/swiftui/view/zindex(_:)). For some reason I wasn't able to make the tooltip appear properly on top of the other views without transparency (despite setting opacity specifically, or trying different alternatives) and using this property is the only thing that fixed it.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift [hunk 2]
```diff
@@ -103,6 +75,28 @@ struct CollapsibleProductRowCard: View {
...
+                        .multilineTextAlignment(.leading)
+                        .lineLimit(nil)
+                        .fixedSize(horizontal: false, vertical: true)
+                        .foregroundColor(.gray)
+                }
+                .frame(maxWidth: .infinity)
+                .padding()
+                .background {
+                    Color.black
+                        .cornerRadius(Layout.frameCornerRadius)
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift]: You should (hopefully) be able to use the `.shadow(.drop)` modifier here to add the drop shadow.

Also... there's some fun to be had adding the little arrow at the top. Seems like a prime candidate for some ChatGPT help if you're not sure how to do it (I'm not sure either, but I think it'd be by combining two other shapes. Figma dev mode suggests using an image of the pointer but that seems like a cop out.)

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift [hunk 3]
```diff
@@ -116,6 +110,44 @@ struct CollapsibleProductRowCard: View {
     }
 }
 
+private extension CollapsibleProductRowCard {
+    @ViewBuilder var discountRow: some View {
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift:174]: Nice one for extracting this.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift [hunk 4]
```diff
@@ -116,6 +110,44 @@ struct CollapsibleProductRowCard: View {
...
+                if let discountLabel = viewModel.discountLabel {
+                    Text(minusSign + discountLabel)
+                        .foregroundColor(.green)
+                }
+            }
+        }
+        Spacer()
+            .renderedIf(!viewModel.hasDiscount)
+        Button {
+            shouldShowCouponsInfoTooltip.toggle()
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift]: For turning it _off_ again, I think we need more than just the button. I instinctively just tapped elsewhere on the screen and that just opens other modals

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/OrderPaymentSection.swift (hunk 1 above)
Comment (joshheald) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/OrderPaymentSection.swift:98]: Nothing to add I'm afraid. Maybe comment about it?

**Review by joshheald — APPROVED** [review]
Added some questions/suggestions. I think we should do them (improve dismissal, give the tooltips more contextual presence by adding the drop shadows and little arrows.)

Once done, and separately from this project, it would be an interesting task to extract the tooltip to be a bit more reusable.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/OrderPaymentSection.swift (hunk 1 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/OrderPaymentSection.swift:98]: Added internal comment: 5122b2c900be724137b3799a4205bdf57f091678

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift (hunk 2 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift]: It seems that property is only available for iOS16 😞 . I've added an item to the [Polish UI issue](https://github.com/woocommerce/woocommerce-ios/issues/10863) so I can explore other modifiers.

> Also... there's some fun to be had adding the little arrow at the top. Seems like a prime candidate for some ChatGPT help if you're not sure how to do it (I'm not sure either, but I think it'd be by combining two other shapes. Figma dev mode suggests using an image of the pointer but that seems like a cop out.)

This was tricky, but I got a decent solution with c41c005a99bc2644babdb60643fcc52d61983215 and bc15d28b6151e8beaaa2b187cc7c0f8685de0243 . It needs more tweaking in order to make it fully reusable but at the moment works for our case. I plan to revisit this shortly, perhaps using ~~[.tooltip](https://developer.apple.com/documentation/uikit/uicontrol/3750832-tooltip)~~ [.popover](https://developer.apple.com/documentation/swiftui/view/popover(ispresented:attachmentanchor:arrowedge:content:)) which seems to be a better solution for what we want, and includes the pointer by default. This is the current tooltip UI:

| Coupons tooltip | Discounts tooltip |
|--------|--------|
| Simulator Screenshot - iPhone 15 - 2023-10-12 at 10 06 55 | Simulator Screenshot - iPhone 15 - 2023-10-11 at 20 07 44 |

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift (hunk 4 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift]: Good catch, addressed in 61f28b59f35f213108d9489780dfbd72485ca27d and b8d3b86938bc9ba8fde244be9ab50b100b0deed7. It's still not perfect, but covers most of the tappable screen.

**Pull request merged**